File: frontend/src/api/platform.ts
import client from './client';

// ── Revisions ────────────────────────────────────────────

export type FileKind =
    | 'manuscript'
    | 'figure'
    | 'supplementary'
    | 'response'
    | 'cover_letter'
    | 'dataset'
    | 'video'
    | 'revised'
    | 'other';

export interface ManuscriptFile {
    id: number;
    kind: FileKind;
    original_filename: string;
    stored_url: string;
    mime_type: string | null;
    size_bytes: number | null;
    created_at: string;
}

export interface ManuscriptVersion {
    id: number;
    submission_id: string;
    version_number: number;
    label: string;
    cover_letter: string | null;
    response_to_reviewers: string | null;
    change_summary: string | null;
    is_current: boolean;
    created_at: string;
    files: ManuscriptFile[];
}

export const fetchVersionsForSubmission = async (
    submissionId: string,
): Promise<ManuscriptVersion[]> => {
    const r = await client.get(`/revisions/submission/${submissionId}`);
    return r.data;
};

export const submitRevision = async (
    submissionId: string,
    payload: {
        label?: string;
        cover_letter?: string;
        response_to_reviewers?: string;
        change_summary?: string;
        files: Array<{
            kind: FileKind;
            original_filename: string;
            stored_url: string;
            mime_type?: string;
            size_bytes?: number;
        }>;
    },
): Promise<ManuscriptVersion> => {
    const r = await client.post(`/revisions/submission/${submissionId}`, payload);
    return r.data;
};

// ── Production ───────────────────────────────────────────

export type ProductionStage =
    | 'copy_editing'
    | 'typesetting'
    | 'proof'
    | 'author_proof_pending'
    | 'author_proof_approved'
    | 'final_pdf'
    | 'doi_assigned'
    | 'published';

export interface ProductionRecord {
    id: number;
    submission_id: string;
    stage: ProductionStage;
    copy_edit_notes: string | null;
    typesetting_notes: string | null;
    proof_pdf_url: string | null;
    author_corrections: string | null;
    final_pdf_url: string | null;
    doi: string | null;
    published_at: string | null;
    updated_at: string;
    created_at: string;
}

export const fetchProductionQueue = async (
    stage?: ProductionStage,
): Promise<ProductionRecord[]> => {
    const r = await client.get('/production/queue', { params: stage ? { stage } : {} });
    return r.data;
};

export const openProductionForSubmission = async (
    submissionId: string,
): Promise<ProductionRecord> => {
    const r = await client.post(`/production/from-accepted/${submissionId}`);
    return r.data;
};

export const updateProduction = async (
    id: number,
    payload: Partial<ProductionRecord>,
): Promise<ProductionRecord> => {
    const r = await client.patch(`/production/${id}`, payload);
    return r.data;
};

// ── Special issues ───────────────────────────────────────

export interface SpecialIssue {
    id: number;
    slug: string;
    title: string;
    description: string;
    guest_editors: string | null;
    topics: string | null;
    cover_image_url: string | null;
    submission_deadline: string | null;
    publication_date: string | null;
    status: 'open' | 'closed' | 'published';
    is_published: boolean;
}

export const fetchSpecialIssues = async (
    include_unpublished = false,
): Promise<SpecialIssue[]> => {
    const r = await client.get('/special-issues/', { params: { include_unpublished } });
    return r.data;
};

export const fetchSpecialIssue = async (slug: string): Promise<SpecialIssue> => {
    const r = await client.get(`/special-issues/${slug}`);
    return r.data;
};

export const createSpecialIssue = async (
    payload: Omit<SpecialIssue, 'id'>,
): Promise<SpecialIssue> => {
    const r = await client.post('/special-issues/', payload);
    return r.data;
};

export const updateSpecialIssue = async (
    slug: string,
    payload: Partial<Omit<SpecialIssue, 'id' | 'slug'>>,
): Promise<SpecialIssue> => {
    const r = await client.patch(`/special-issues/${slug}`, payload);
    return r.data;
};

export const deleteSpecialIssue = async (slug: string): Promise<void> => {
    await client.delete(`/special-issues/${slug}`);
};

// ── Email templates ──────────────────────────────────────

export interface EmailTemplate {
    id: number;
    slug: string;
    subject: string;
    body: string;
    description: string | null;
    placeholders: string | null;
    is_active: boolean;
    updated_by: string | null;
    updated_at: string;
}

export const fetchEmailTemplates = async (): Promise<EmailTemplate[]> => {
    const r = await client.get('/email-templates/');
    return r.data;
};

export const updateEmailTemplate = async (
    slug: string,
    payload: Partial<Omit<EmailTemplate, 'id' | 'slug' | 'updated_by' | 'updated_at'>>,
): Promise<EmailTemplate> => {
    const r = await client.patch(`/email-templates/${slug}`, payload);
    return r.data;
};

// ── Audit log ────────────────────────────────────────────

export interface AuditLogEntry {
    id: number;
    actor_id: number | null;
    actor_email: string | null;
    action: string;
    target_type: string | null;
    target_id: string | null;
    ip_address: string | null;
    meta: Record<string, unknown> | null;
    created_at: string;
}

export const fetchAuditLog = async (params?: {
    action?: string;
    actor_email?: string;
    target_type?: string;
    q?: string;
    limit?: number;
}): Promise<AuditLogEntry[]> => {
    const r = await client.get('/audit-logs/', { params });
    return r.data;
};

// ── References ───────────────────────────────────────────

export interface ArticleReference {
    id: number;
    article_id: number;
    sequence: number;
    text: string;
    doi: string | null;
    url: string | null;
}

export const fetchReferences = async (articleId: number): Promise<ArticleReference[]> => {
    const r = await client.get(`/references/article/${articleId}`);
    return r.data;
};

export const addReference = async (
    articleId: number,
    payload: { sequence?: number; text: string; doi?: string; url?: string },
): Promise<ArticleReference> => {
    const r = await client.post(`/references/article/${articleId}`, payload);
    return r.data;
};

export const deleteReference = async (referenceId: number): Promise<void> => {
    await client.delete(`/references/${referenceId}`);
};

// ── Users admin ──────────────────────────────────────────

export type AdminRole = 'author' | 'editor' | 'section_editor' | 'admin';

export interface AdminUser {
    id: number;
    username: string | null;
    email: string;
    full_name: string | null;
    role: AdminRole;
    is_active: boolean;
    country: string | null;
    institution: string | null;
    orcid: string | null;
    mfa_enabled: boolean;
}

export const fetchAdminUsers = async (params?: {
    role?: AdminRole;
    q?: string;
}): Promise<AdminUser[]> => {
    const r = await client.get('/users-admin/', { params });
    return r.data;
};

export const updateAdminUser = async (
    id: number,
    payload: Partial<Omit<AdminUser, 'id' | 'email' | 'username' | 'orcid' | 'mfa_enabled'>>,
): Promise<AdminUser> => {
    const r = await client.patch(`/users-admin/${id}`, payload);
    return r.data;
};

export const deactivateAdminUser = async (id: number): Promise<AdminUser> => {
    const r = await client.post(`/users-admin/${id}/deactivate`);
    return r.data;
};

File: backend/app/schemas/platform.py
"""Shared Pydantic schemas for the platform-expansion routers.

Grouped in one module to keep the router imports light. Individual routers
import only what they need.
"""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


# ─── Manuscript version + files ──────────────────────────

class ManuscriptFileRead(BaseModel):
    id: int
    kind: str
    original_filename: str
    stored_url: str
    mime_type: Optional[str] = None
    size_bytes: Optional[int] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ManuscriptFileCreate(BaseModel):
    kind: str = Field(pattern="^(manuscript|figure|supplementary|response|cover_letter|dataset|video|revised|other)$")
    original_filename: str = Field(min_length=1, max_length=400)
    stored_url: str = Field(min_length=1, max_length=1024)
    mime_type: Optional[str] = None
    size_bytes: Optional[int] = None
    checksum: Optional[str] = None


class ManuscriptVersionRead(BaseModel):
    id: int
    submission_id: str
    version_number: int
    label: str
    cover_letter: Optional[str] = None
    response_to_reviewers: Optional[str] = None
    change_summary: Optional[str] = None
    is_current: bool
    created_at: datetime
    files: List[ManuscriptFileRead] = []
    model_config = ConfigDict(from_attributes=True)


class ManuscriptVersionCreate(BaseModel):
    label: Optional[str] = Field(default=None, max_length=80)
    cover_letter: Optional[str] = None
    response_to_reviewers: Optional[str] = None
    change_summary: Optional[str] = None
    files: List[ManuscriptFileCreate] = []


# ─── Production ──────────────────────────────────────────

class ProductionRead(BaseModel):
    id: int
    submission_id: str
    stage: str
    copy_edit_notes: Optional[str] = None
    typesetting_notes: Optional[str] = None
    proof_pdf_url: Optional[str] = None
    author_corrections: Optional[str] = None
    final_pdf_url: Optional[str] = None
    doi: Optional[str] = None
    published_at: Optional[datetime] = None
    updated_at: datetime
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ProductionUpdate(BaseModel):
    stage: Optional[str] = Field(default=None, pattern="^(copy_editing|typesetting|proof|author_proof_pending|author_proof_approved|final_pdf|doi_assigned|published)$")
    copy_edit_notes: Optional[str] = None
    typesetting_notes: Optional[str] = None
    proof_pdf_url: Optional[str] = None
    author_corrections: Optional[str] = None
    final_pdf_url: Optional[str] = None
    doi: Optional[str] = None
    published_at: Optional[datetime] = None


# ─── Special issues ──────────────────────────────────────

class SpecialIssueBase(BaseModel):
    slug: str = Field(min_length=1, max_length=120)
    title: str = Field(min_length=1, max_length=300)
    description: str = Field(min_length=1)
    guest_editors: Optional[str] = None
    topics: Optional[str] = None
    cover_image_url: Optional[str] = None
    submission_deadline: Optional[datetime] = None
    publication_date: Optional[datetime] = None
    status: str = Field(default="open", pattern="^(open|closed|published)$")
    is_published: bool = True


class SpecialIssueCreate(SpecialIssueBase):
    pass


class SpecialIssueUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    guest_editors: Optional[str] = None
    topics: Optional[str] = None
    cover_image_url: Optional[str] = None
    submission_deadline: Optional[datetime] = None
    publication_date: Optional[datetime] = None
    status: Optional[str] = Field(default=None, pattern="^(open|closed|published)$")
    is_published: Optional[bool] = None


class SpecialIssueRead(SpecialIssueBase):
    id: int
    model_config = ConfigDict(from_attributes=True)


# ─── Email templates ─────────────────────────────────────

class EmailTemplateBase(BaseModel):
    slug: str = Field(min_length=1, max_length=80)
    subject: str = Field(min_length=1, max_length=300)
    body: str = Field(min_length=1)
    description: Optional[str] = None
    placeholders: Optional[str] = None
    is_active: bool = True


class EmailTemplateCreate(EmailTemplateBase):
    pass


class EmailTemplateUpdate(BaseModel):
    subject: Optional[str] = None
    body: Optional[str] = None
    description: Optional[str] = None
    placeholders: Optional[str] = None
    is_active: Optional[bool] = None


class EmailTemplateRead(EmailTemplateBase):
    id: int
    updated_by: Optional[str] = None
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ─── Audit log ───────────────────────────────────────────

class AuditLogRead(BaseModel):
    id: int
    actor_id: Optional[int] = None
    actor_email: Optional[str] = None
    action: str
    target_type: Optional[str] = None
    target_id: Optional[str] = None
    ip_address: Optional[str] = None
    meta: Optional[dict] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ─── Article references ─────────────────────────────────

class ArticleReferenceCreate(BaseModel):
    sequence: int = 1
    text: str = Field(min_length=1)
    doi: Optional[str] = None
    url: Optional[str] = None


class ArticleReferenceRead(ArticleReferenceCreate):
    id: int
    article_id: int
    model_config = ConfigDict(from_attributes=True)


# ─── User admin ──────────────────────────────────────────

class UserAdminRead(BaseModel):
    id: int
    username: Optional[str] = None
    email: str
    full_name: Optional[str] = None
    role: str
    is_active: bool
    country: Optional[str] = None
    institution: Optional[str] = None
    orcid: Optional[str] = None
    mfa_enabled: bool = False
    model_config = ConfigDict(from_attributes=True)


class UserAdminUpdate(BaseModel):
    full_name: Optional[str] = None
    role: Optional[str] = Field(default=None, pattern="^(author|editor|section_editor|admin)$")
    is_active: Optional[bool] = None
    country: Optional[str] = None
    institution: Optional[str] = None
